std/digest/xxh.d

```diff
@@ -0,0 +1,3352 @@
+ * alignment on 32-bit. This isn't enough for the 32 byte aligned loads in AVX2
+ * or on 32-bit, the 16 byte aligned loads in SSE2 and NEON.
+ *
+ * This underalignment previously caused a rather obvious crash which went
+ * completely unnoticed due to XXH3_createState() not actually being tested.
+ * Credit to RedSpah for noticing this bug.
+ *
+ * The alignment is done manually: Functions like posix_memalign or _mm_malloc
+ * are avoided: To maintain portability, we would have to write a fallback
+ * like this anyways, and besides, testing for the existence of library
+ * functions without relying on external build tools is impossible.
+ *
+ * The method is simple: Overallocate, manually align, and store the offset
+ * to the original behind the returned pointer.
+ *
+ * Align must be a power of 2 and 8 <= align <= 128.
+ */
+private void* XXH_alignedMalloc(size_t s, size_t align_) @trusted nothrow @nogc
+{
+    import core.stdc.stdlib : malloc;
+
+    assert(align_ <= 128 && align_ >= 8, "align out of range"); /* range check */
+    assert((align_ & (align_ - 1)) == 0, "(align_ & (align_ - 1)) != 0"); /* power of 2 */
+    assert(s != 0 && s < (s + align_), "s == 0 || s >= (s + align_)"); /* empty/overflow */
+    { /* Overallocate to make room for manual realignment and an offset byte */
+        xxh_u8* base = cast(xxh_u8*) malloc(s + align_);
+        if (base != null)
+        {
+            /*
+             * Get the offset needed to align this pointer.
+             *
+             * Even if the returned pointer is aligned, there will always be
+             * at least one byte to store the offset to the original pointer.
+             */
+            size_t offset = align_ - (cast(size_t) base & (align_ - 1)); /* base % align */
+            /* Add the offset for the now-aligned pointer */
+            xxh_u8* ptr = base + offset;
+
+            assert(cast(size_t) ptr % align_ == 0, "ptr % align_ != 0");
+
+            /* Store the offset immediately before the returned pointer. */
+            ptr[-1] = cast(xxh_u8) offset;
+            return ptr;
+        }
+        return null;
+    }
+}
+/*
+ * Frees an aligned pointer allocated by XXH_alignedMalloc(). Don't pass
+ * normal malloc'd pointers, XXH_alignedMalloc has a specific data layout.
+ */
+private void XXH_alignedFree(void* p) @trusted nothrow @nogc
+{
+    import core.stdc.stdlib : free;
+
+    if (p != null)
+    {
+        xxh_u8* ptr = cast(xxh_u8*) p;
+        /* Get the offset byte we added in XXH_malloc. */
+        xxh_u8 offset = ptr[-1];
+        /* Free the original malloc'd pointer */
+        xxh_u8* base = ptr - offset;
+        free(base);
+    }
+}
+
+private void XXH3_INITSTATE(XXH3_state_t* XXH3_state_ptr) @trusted nothrow @nogc
+{
+    (XXH3_state_ptr).seed = 0;
+}
+
+/*! @ingroup XXH3_family */
+XXH3_state_t* XXH3_createState() @trusted nothrow @nogc
+{
+    XXH3_state_t* state = cast(XXH3_state_t*) XXH_alignedMalloc((XXH3_state_t).sizeof, 64);
```

Comment:
The alignment of the struct can be fetched with `StructType.alignof`.

Reply:
Ah, thanks. That's better than this magic number.

Reply:
Interface dropped.